appengine/test/tc_sample_apps.rb

```diff
@@ -0,0 +1,49 @@
+require "minitest/autorun"
+require_relative "test_helper.rb"
+
+
+# Tests of a bunch of sample apps. Treats every subdirectory of the test
+# directory that contains a Dockerfile as a sample app. Builds the docker
+# image, runs the server, and hits the root of the server, expecting the
+# string "ruby app" to be returned.
+#
+# This is supposed to exercise the base image extended in various ways, so
+# the sample app Dockerfiles should all inherit FROM the base image, which
+# will be called "appengine-ruby-base".
+
+class TestSampleApps < ::Minitest::Test
+
+  include TestHelper
+
+
+  ::Dir.glob("#{::File.dirname(__FILE__)}/*/Dockerfile").each do |dockerfile|
+    test_dir = ::File.dirname(dockerfile)
+    base_name = ::File.basename(test_dir)
+    define_method("test_#{base_name}") do
+      run_app_test(base_name)
+    end
+  end
+
+
+  def run_app_test(dirname)
+    puts("**** Testing app: #{dirname}")
+    ::Dir.chdir(::File.join(::File.dirname(__FILE__), dirname)) do |dir|
+      num = rand(1000000)
```

Comment:
It's there only to dedup the image and container names in case the tests happen to be run concurrently. Now that I think about it, it's simpler just to use the dirname. Changed accordingly.